ironic-lib/0004-Filesystem-creation-implementation.patch

```diff
@@ -0,0 +1,249 @@
+From 92a4f9e326ab9f5285d99b490da32c05981c2068 Mon Sep 17 00:00:00 2001
+From: "Grzegorz Grasza (xek)" <grzegorz.grasza@intel.com>
+Date: Thu, 11 Jan 2018 16:56:01 +0100
+Subject: [PATCH 4/8] Filesystem creation implementation
+
+Co-Authored-By: Grzegorz Grasza <grzegorz.grasza@intel.com>
+---
+ ironic_lib/system_installer/filesystemsetup.py | 70 ++++++++++++++++++-
+ ironic_lib/tests/examples/example_efi.yaml     | 23 +++++++
+ ironic_lib/tests/test_examples.py              |  4 ++
+ ironic_lib/tests/test_filesystemsetup.py       | 95 ++++++++++++++++++++++++++
+ 4 files changed, 190 insertions(+), 2 deletions(-)
+ create mode 100644 ironic_lib/tests/examples/example_efi.yaml
+ create mode 100644 ironic_lib/tests/test_filesystemsetup.py
+
+diff --git a/ironic_lib/system_installer/filesystemsetup.py b/ironic_lib/system_installer/filesystemsetup.py
+index 3729298..92270a7 100644
+--- a/ironic_lib/system_installer/filesystemsetup.py
++++ b/ironic_lib/system_installer/filesystemsetup.py
+@@ -13,7 +13,15 @@
+ # See the License for the specific language governing permissions and
+ # limitations under the License.
+ 
++import os
++
++from ironic_lib import utils
++from oslo_log import log
++
+ from ironic_lib.system_installer.base import Setup
++from ironic_lib.system_installer import exceptions
++
++LOG = log.getLogger()
+ 
+ 
+ class FilesystemSetup(Setup):
+@@ -24,8 +32,66 @@ class FilesystemSetup(Setup):
+     def get_src_names(self):
+         return list(self.conf)
+ 
++    def validate_conf(self):
++        if not self.get_boot_partition():
++            raise exceptions.ConfError(
++                'No boot partition found in provided config')
++
++    def validate_env(self):
++        for disk in self.conf.values():
++            fstype = disk.get('fstype', 'xfs')
++            if fstype != 'swap' and not self._cmd_exists('mkfs.' + fstype):
++                raise exceptions.EnvError('No mkfs.' + fstype)
++
++    def setup_disks(self, devices):
++        for disk_name, disk in self.conf.items():
++            fstype = disk.get('fstype', 'ext4')
++            device = devices.get(disk_name, disk_name)
++            if not disk.get('preserve', False):
++                if fstype == 'swap':
++                    out, err = utils.execute('mkswap', device)
++                else:
++                    out, err = utils.execute('yes | mkfs.{} {}'.format(
++                        fstype, device), shell=True)
++                LOG.debug("mkfs stdout: {}".format(out))
++                LOG.debug("mkfs stderr: {}".format(err))
```

Comment:
if there is an error in mkfs, we should stop the installation right?

Reply:
Yes, an exception is raised by utils.execute in this case.